src/OpenCensus/Resource/Resource.cs

```diff
@@ -0,0 +1,202 @@
+            }
+            catch (SecurityException)
+            {
+                openCensusResourceType = GlobalResourceType;
+
+                // TODO - Log error
+            }
+
+            try
+            {
+                openCensusEnvironmentTags = Environment.GetEnvironmentVariable(OC_RESOURCE_LABELS_ENV);
+            }
+            catch (SecurityException)
+            {
+                openCensusEnvironmentTags = string.Empty;
+
+                // TODO - Log error
+            }
+
+            TryParseResourceType(openCensusResourceType, out EnvironmentType);
+            EnvironmentToLabelMap = ParseResourceLabels(Environment.GetEnvironmentVariable(OC_RESOURCE_LABELS_ENV));
+        }
+
+        /// <summary>
+        /// Gets the identification of the resource.
+        /// </summary>
+        public abstract string Type { get; protected set; }
+
+        /// <summary>
+        /// Gets the map between the tag and its value.
+        /// </summary>
+        public abstract IList<ITag> Tags { get; }
+
+        /// <summary>
+        /// Creates a label/tag map from the OC_RESOURCE_LABELS environment variable.
+        /// OC_RESOURCE_LABELS: A comma-separated list of labels describing the source in more detail,
+        /// e.g. “key1=val1,key2=val2”. Domain names and paths are accepted as label keys.
+        /// Values may be quoted or unquoted in general. If a value contains whitespaces, =, or " characters, it must
+        /// always be quoted.
+        /// </summary>
+        /// <param name="rawEnvironmentTags">Environment tags as a raw, comma separated string</param>
+        /// <returns>Environment Tags as a list</returns>
+        private static ITag[] ParseResourceLabels(string rawEnvironmentTags)
+        {
+            if (rawEnvironmentTags == null)
+            {
+                return new ITag[0] { };
+            }
+            else
+            {
+                var labels = new List<ITag>();
+                string[] rawLabels = rawEnvironmentTags.Split(LABEL_LIST_SPLITTER);
+
+                Regex regex = new Regex("^\"|\"$", RegexOptions.Compiled);
+                
+                foreach (var rawLabel in rawLabels)
+                {
+                    string[] keyValuePair = rawLabel.Split(LABEL_KEY_VALUE_SPLITTER);
+                    if (keyValuePair.Length != 2)
+                    {
+                        continue;
+                    }
+
+                    string key = keyValuePair[0].Trim();
+                    string value = Regex.Replace(keyValuePair[1].Trim(), "^\"|\"$", string.Empty);
+
+                    Arguments.Check(IsValidAndNotEmpty(key), "Label key" + ERROR_MESSAGE_INVALID_CHARS);
+                    Arguments.Check(IsValid(value), "Label value" + ERROR_MESSAGE_INVALID_CHARS);
+
+                    labels.Add(new Tag(new TagKey(key), new TagValue(value)));
+                }
+
+                return labels.ToArray();
```

Comment:
this is an extra allocation, could we avoid it by using same type everywhere?

Reply:
If you mean TagKey and TagValue constructors - I've fixed that. If you mean labels.ToArray call - there is not much you can do. The number of labels is not known in advance so you have to parse them in runtime. I could return a list but once labels were calculated, they should stay immutable. Finally, this calculation happens once in a lifetime of the process.